compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -1576,6 +1576,45 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         false
     }
 
+    /// When a failed try operator follows an `unwrap()` or `expect()`, suggest removing it
+    fn suggest_remove_unwrap(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diag<'_>,
+    ) -> bool {
+        if !matches!(obligation.cause.code(), ObligationCauseCode::QuestionMark) {
+            return false;
+        }
+
+        let Some(body) = self.tcx.hir().maybe_body_owned_by(obligation.cause.body_id) else {
+            return false;
+        };
+
+        let mut expr_finder = super::FindExprBySpan::new(obligation.cause.span, self.tcx);
+        expr_finder.visit_expr(body.value);
+
+        if let Some(expr) = expr_finder.result
+            && let Some(node) = self.tcx.hir().parent_iter(expr.hir_id).nth(2)
+            && let hir::Node::Expr(expr) = node.1
+            && let hir::ExprKind::Match(expr, _, _) = expr.kind
+            && let hir::ExprKind::Call(_, args) = expr.kind
+            && let Some(expr) = args.get(0)
+            && let hir::ExprKind::MethodCall(pathsegment, receiver, _, span) = expr.kind
+            && (pathsegment.ident.name == sym::unwrap || pathsegment.ident.name == sym::expect)
+        {
+            err.span_suggestion_verbose(
+                receiver.span.shrink_to_hi().to(span),
+                format!("remove the `.{}()`", pathsegment.ident.name),
+                "",
+                Applicability::MaybeIncorrect,
+            );
```

Comment:
Before suggesting, also check that removing the `unwrap()` actually makes the `?` not fail. i.e. check that the predicate holds with the type unwrapped